src/ResultSet/AbstractResultSet.php

```diff
@@ -208,7 +203,7 @@ public function current()
         if (is_array($this->buffer)) {
             $this->buffer[$this->position] = $data;
         }
-        return is_array($data) ? $data : null;
+        return (is_array($data) || $data instanceof Countable) ? $data : null;
```

Comment:
Why is `Countable` a valid type, but (for example) `Traversable` not? What's the rationale behind the change?

Reply:
Since this apparently fixes #330 because of a bug I might have introduced earlier, it is not needed to cover every single interface that could be returned by ->current(). If it fixes #330 and does not break anything, then it is ok, but we should see a unit test that confirms #330 and thus explains the need for the fix.

Reply:
@bartmcleod The code that you introduced [here](https://github.com/zendframework/zend-db/blob/8276cf6cfe45dce834979f0e714eea81065a7493/src/ResultSet/AbstractResultSet.php#L197-L200) just returns the current value of `dataSource` that can be a different type, depending on the adapter (e.g. for Mysqli is [false](https://github.com/zendframework/zend-db/blob/master/src/Adapter/Driver/Mysqli/Result.php#L56), for PDO is [null](https://github.com/zendframework/zend-db/blob/master/src/Adapter/Driver/Pdo/Result.php#L53), etc). Maybe, I should align all the default value of the adapter to be the same but I'm not sure this will introduce BC breaks in some existing project. That's why I focused my fix on the `AbstractResultSet`.

@Ocramius I'm trying to accomodate different return type for `Resultset::current()` coming from different adapter, it's seems there is inconsistency with different adapters (see above). Anyway, I'm still testing all the use cases, it's quite difficult without type hinting...

Reply:
@ezimuel would it be okay to check with `empty()`  and to always return null in that case? So that false and null are treated equally.